### ISSUE TITLE
Consider exposing consts for some set of common locales

### ISSUE THREAD

**Issue by cmyr** [zbraniecki/unic-locale#10]
Not sure if you think this would make sense? It may be of limited utility in actual use, but would maybe be helpful when constructing fallback chains, without needing to bring in the macro.

**Comment by zbraniecki** [comment]
I've seen that ICU does it. I'm not sure if I like that from both technical, and i18n (sic!) perspective.

Can you describe why using the `langid!("en")` is a burden that `StaticLocales::EN` solve?

**Comment by cmyr** [comment]
using `langid!` requires me to bring in `syn` and co, which has a non-negligible impact on compile time and possibly code size (I haven't measured the the code size, I just see that frequently with proc macros, so this may be an unfair generalization).

I do think proc macros have their place, but when there are reasonable alternatives I prefer that.

When [RFC 2345](https://github.com/rust-lang/rust/issues/51999) lands it will potentially offer a nice elegant where you can have `const` constructors that validate their arguments at compile time. (this also might require [`const_str_as_bytes`](https://github.com/rust-lang/rust/issues/57563))?

**Comment by zbraniecki** [comment]
I definitely am looking forward to be able to use langid macro with constants! We have `from_uncheded_raw_parts` - which should handle that.

**Comment by cmyr** [comment]
What I was imagining is you could have constructors that take `&'static str` arguments, and can verify that they're the correct length. Unless I'm mistaken, `from_unchecked_raw_parts` requires you to generate some integer representation of your strings, which feels a bit scary (although I guess you could do something like the following?)

```rust
const fn simple_lang_id(lang: (char, char)) -> LanguageIdentifier {
    let raw = (lang.1 as u64) << 8 | (lang.0 as u64);
    unsafe {
        let lang =  TinyStr8::new_unchecked(raw);
        LanguageIdentifier::from_raw_parts_unchecked(Some(lang), None, None, None)
    }
}

const EN: LanguageIdentifier = simple_lang_id(('e', 'n'));
```

Anyway I'm mostly just playing around, feel free to ignore me. 😅💦

**Comment by cmyr** [comment]
Not sure if this is useful to you, but [here are some numbers](https://github.com/xi-editor/druid/pull/124#issuecomment-523211377) on the impact of bringing fluent & unic-locale into druid. Some of this looks to be rental, but about 100kb of additional executable size seems to be from the `lang_id!` macro, which is brought in by `fluent-bundle`; so I think there's a reasonable case that a few `consts` would be a good alternative.

Obviously we don't want to be too imperialistic, but would you accept a PR that adds, say, [ten  common locale identifiers](https://stackoverflow.com/questions/9711066/most-common-locales-for-worldwide-compatibility/9713377#9713377) as consts?

**Comment by zbraniecki** [comment]
So, my goal was to get to the point where one can do:

```rust
const EN = langid!("en");
```

and it should compile into:

```rust
const EN = LanguageIdentifier::from_raw_parts_unchecked(TinyStr8::new_unchecked(25956), None, TinyStr4::new_unchecked(17732), None);
```

Is that not going to fly?

**Comment by zbraniecki** [comment]
> Not sure if this is useful to you, but here are some numbers on the impact of bringing fluent & unic-locale into druid. Some of this looks to be rental, but about 100kb of additional executable size seems to be from the lang_id! macro, which is brought in by fluent-bundle; so I think there's a reasonable case that a few consts would be a good alternative.

Yeah, that's definitely a good reason to remove the macro from fluent-bundle.

> Obviously we don't want to be too imperialistic, but would you accept a PR that adds, say, ten common locale identifiers as consts?

I wouldn't. I would cringe a little bit and hope there's a better solution, but end up being pragmatic.

But... how would those 10 save you? What use case you have where static 10 locales allow you to skip dynamically producing them?

**Comment by cmyr** [comment]
`langid!` producing consts would be good, but it doesn't fix the fundamental problem that proc macros are expensive, and should, in my opinion, be reserved for cases where there are no reasonable alternatives.

My use case is generating a default locale for fallback, and potentially for tests. This seems to be mainly what `langid!` is for? For default locales, I suspect there are a small set of locales that are the default in the majority of situations, and if those were exposed as `consts` in this crate that would cover the majority of the current use of `langid!`.

Realistically, those locales might just be `en` and `en-US`, but that feels colonial, so if I were going to add any locales I would prefer to pick a _slightly_ larger set, guided by something like the link above.

**Comment by cmyr** [comment]
> > Obviously we don't want to be too imperialistic, but would you accept a PR that adds, say, ten common locale identifiers as consts?
> 
> I wouldn't. I would cringe a little bit and hope there's a better solution, but end up being pragmatic.

Also slightly confused here: the second sentence implies to me that you _would_ consider this PR?

**Comment by cmyr** [comment]
Okay, ran into another hiccup: Is there any way to get a `Box<[TinyStr8]>` in a const context? If there is I can't figure it out, which means i can't use `from_raw_parts_unchecked` in a const context 😒.

**Comment by raphlinus** [comment]
I know we're talking about const fn's a lot, but in my experimentation on `TinyStr`, it is often the case that simple inlining does the right thing. In particular, `TinyStr4::parse("en").unwrap()` does in fact compile to `25956`. It's possible you could carefully craft the inlines so that a plain function call (no macros) with a fixed string would yield efficient code.

Also in my experimentation, `const` is extremely limited right now and basically can't do any control flow. I think it's on the roadmap for that to change. So it's possible the function approach would be "good enough" now and would become const-able later.

I don't have the answers here, but think it might be worth exploring more.

**Comment by zbraniecki** [comment]
> My use case is generating a default locale for fallback, and potentially for tests. 

Ok, so what we're looking for is a way to say:

"here's a string that I want to guarantee is a well-formed language identifier at build time, and pay zero cost for it at runtime", right?

I think I'm misunderstanding how proc_macro work, but I expected it to give us exactly that.

Your proposal is to expose some `LanguageIdentifier::EN` which would be guaranteed to be an `en-US` language identifier.
I'm reluctant to go this route because hardcoding a subset of locales feels awkward (any cutting point will be arbitrary), but if we can't find any better way, then I'm ok accepting this route.

My litmus test is - you say "I need it for fallback" - but who is there to decide what should be the fallback locale? If I hardcode `en`, and someone wants `sr` as a fallback because it's their source locale, how do we avoid a bias toward `en`?

@Manishearth - do you have any ideas how can we solve the problem?

**Comment by zbraniecki** [comment]
> Also slightly confused here: the second sentence implies to me that you would consider this PR?

Yes, I'd accept it if we can find no other way to achieve it without hardcoding. :)

**Comment by zbraniecki** [comment]
> Okay, ran into another hiccup: Is there any way to get a Box<[TinyStr8]> in a const context? If there is I can't figure it out, which means i can't use from_raw_parts_unchecked in a const context 😒.

Yep, we're waiting for more const stabilizations!

**Comment by cmyr** [comment]
Yea, we would need `LanguageIdentifier` to have a `const`-able constructor to move ahead with this. This would be possible if `variants` was an `Option`, but I don't know if this makes sense.

It's also very possible that I'm misidentifying `langid!` as being particularly to blame; it looks like it isn't adding much code at all in a release build. It does impact compile times, though, if you otherwise don't need `syn` and `proc_macro`.

**Comment by zbraniecki** [comment]
> This would be possible if variants was an Option, but I don't know if this makes sense.

 I was considering it, my `from_parts` even accepts an Option.

The challenge is that in reality it wouldn't make sense I think because it would just make macros `constable` only if there are no variants in the locale. It feel weird that we'd have functionality that works for a subset of well formed locales, but then, since variants are rare, maybe its an ok tradeoff while we wait for Rust to bring more const?

> It does impact compile times, though, if you otherwise don't need syn and proc_macro.

Gotcha, I'm interested in reducing that as well, just as a separate problem from inspecting what impacts size and reducing that.

**Comment by cmyr** [comment]
A way that I like to think about things like this: if I can write an elegant API that serves 95% of cases, I'm happy to have an inelegant backup API for folks that really need it, and I think that the duplication is okay.

So: I think it's totally okay if you can only use `const` when there are no variants; and I think it's okay if your api for this is something like,

```rust
pub unsafe const fn from_lang_and_region_unchecked(lang: TinyStr8, region: TinyStr4) -> LanguageIdentifier
```

**Comment by zbraniecki** [comment]
Would it remove your need for hardcoded langids?

**Comment by zbraniecki** [comment]
@cmyr - ping! I'm getting ready to release the next unic-locale with the switch to TinyStr. Would it be beneficial for you if we made variants an Option and added a macro that handles only language/script/region and panics on variants, but works on static?

**Comment by cmyr** [comment]
I think it would be good to have variants be an Option, and to have a `const` constructor that just handles language/script/region. For macro, are you thinking of a `macro_rules!` macro? I could see that being helpful, but I'd have to see what you have in mind. With the `const` constructor I can at least get the consts I want, even if the code is ugly, and that's fine by me; I can make my own wrappers etc as required.

**Comment by zbraniecki** [comment]
My thinking was to add a proc macro similar to langid - const_langid which would still validate but produce something that can handle const and panic on variant.

**Comment by cmyr** [comment]
I wonder if you could have the same macro call the `const` constructor when possible, and the non-`const` otherwise?

In any case that sounds good. I would also like to be able to experiment with the `const` constructor without using the macro, so as long as it's public that's great.  👍

**Comment by zbraniecki** [comment]
@cmyr - just released 0.5.0 with those changes!

**Comment by cmyr** [comment]
@zbraniecki Thank you!
